File: go/heap/heap.go
package heap

import . "golang.org/x/exp/constraints"

// MinHeap is a struct representing a MinHeap
type MinHeap[T Ordered] struct {
	length int
	data   []T
}

// NewMinHeap is a constructor function for a MinHeap, it takes in a function which decides which compares 2 values of type T
func NewMinHeap[T Ordered]() *MinHeap[T] {
	return &MinHeap[T]{
		length: 0,
		data:   []T{},
	}
}

// Length returns the number of items in the heap
func (heap *MinHeap[T]) Length() int {
	return heap.length
}

// Insert inserts a new value into the MinHeap
func (heap *MinHeap[T]) Insert(value T) {
	heap.data = append(heap.data, value)
	heap.heapifyUp(heap.length)
	heap.length++
}

// Delete deletes the first value in the MinHeap
func (heap *MinHeap[T]) Delete() (T, error) {
	if heap.length == 0 {
		return *new(T), ErrDeleteFromEmptyHeap
	}

	out := heap.data[0]
	heap.length--

	if heap.length == 0 {
		heap.data = []T{}
		return out, nil
	}

	heap.data[0] = heap.data[heap.length]
	heap.data = heap.data[:heap.length]
	heap.heapifyDown(0)

	return out, nil
}

// heapifyUp starts at the end of the heap and recursively bubbles the smallest value inside the heap to the top
func (heap *MinHeap[T]) heapifyUp(idx int) {
	if idx == 0 {
		return
	}

	pIdx := parent(idx)
	value := heap.data[idx]
	parentValue := heap.data[pIdx]

	if parentValue > value {
		heap.data[idx] = parentValue
		heap.data[pIdx] = value
		heap.heapifyUp(pIdx)
	}
}

// heapifyDown starts at the top of the heap and recursively bubbles a value to its appropriate position inside the heap
func (heap *MinHeap[T]) heapifyDown(idx int) {
	lIdx := leftChild(idx)
	rIdx := rightChild(idx)

	if idx >= heap.length || lIdx >= heap.length {
		return
	}

	value := heap.data[idx]
	lValue := heap.data[lIdx]

	if rIdx == heap.length {
		if value > lValue {
			heap.data[idx] = lValue
			heap.data[lIdx] = value
		}
		return
	}

	rValue := heap.data[rIdx]

	if lValue > rValue && value > rValue {
		heap.data[idx] = rValue
		heap.data[rIdx] = value
		heap.heapifyDown(rIdx)
	} else if value > lValue {
		heap.data[idx] = lValue
		heap.data[lIdx] = value
		heap.heapifyDown(lIdx)
	}
}

File: ts/heap/heap.js
export class Heap {
    length = 0;
    data = [];

    peek() {
        if (!this.length) return undefined;
        return this.data[0];
    }

    add(value) {
        this.data.push(value);
        this.heapifyUp(this.length);
        this.length++;
    }

    delete() {
        if (!this.length) return undefined;

        this.length--;
        if (!this.length) return this.data.pop();

        const out = this.data[0];
        this.data[0] = this.data.pop();
        this.heapifyDown(0);

        return out;
    }

    heapifyUp(idx) {
        if (!idx) return;

        const parentIdx = this.parent(idx);
        const value = this.data[idx];
        const parentValue = this.data[parentIdx];

        if (
            (Array.isArray(value) && value[0] < parentValue[0]) ||
            (!Array.isArray(value) && value < parentValue)
        ) {
            this.data[idx] = parentValue;
            this.data[parentIdx] = value;
            this.heapifyUp(parentIdx);
        }
    }

    heapifyDown(idx) {
        const leftIdx = this.leftChild(idx);
        const rightIdx = this.rightChild(idx);

        if (idx >= this.length || leftIdx >= this.length) return;

        const value = this.data[idx];
        const leftValue = this.data[leftIdx];

        if (rightIdx === this.length) {
            if (
                (Array.isArray(value) && value[0] > leftValue[0]) ||
                (!Array.isArray(value) && value > leftValue)
            ) {
                this.data[idx] = leftValue;
                this.data[leftIdx] = value;
            }
            return;
        }

        const rightValue = this.data[rightIdx];

        if (
            (Array.isArray(value) &&
                leftValue[0] > rightValue[0] &&
                value[0] > rightValue[0]) ||
            (!Array.isArray(value) &&
                leftValue > rightValue &&
                value > rightValue)
        ) {
            this.data[idx] = rightValue;
            this.data[rightIdx] = value;
            this.heapifyDown(rightIdx);
        } else if (
            (Array.isArray(value) && value[0] > leftValue[0]) ||
            (!Array.isArray(value) && value > leftValue)
        ) {
            this.data[idx] = leftValue;
            this.data[leftIdx] = value;
            this.heapifyDown(leftIdx);
        }
    }

    parent(idx) {
        return Math.floor((idx - 1) / 2);
    }
    leftChild(idx) {
        return idx * 2 + 1;
    }
    rightChild(idx) {
        return idx * 2 + 2;
    }
}
